### ISSUE TITLE
Cannot combine gcc built libraries with builds using  intel-oneapi

### ISSUE THREAD

**Issue by jcwright77** [LLNL/spack#47319]
### Steps to reproduce

``` 
module load gcc/12.2.2
module load sed
module load intel-oneapi-compilers
```

```
Lmod is automatically replacing "gcc/12.2.0" with "intel-oneapi-compilers/2024.1.0".


Inactive Modules:
  1) sed

Due to MODULEPATH changes, the following have been reloaded:
  1) glibc/2.28
```

So if I want to build something with the oneapi compiler and use a gcc built library (say for a graphic library or some system library), I can't? Is there a proper way to do this or does everything have to be rebuilt with the compiler being used.

I swear this worked before the introduction of gcc runtimes.

### Error message

space debug is not applicable here. It has to do with modules space generates. Probably the cause for the behavior is the line

```family('compiler')``` in the module (lmod) files

### Information on your system

* **Spack:** 0.22.0 (5fe93fee1eec46a0750bd340198bffcb92ff9eec)
* **Python:** 3.6.8
* **Platform:** linux-rocky8-broadwell
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by haampie** [comment]
[Exclude `glibc` from the modules](https://spack.readthedocs.io/en/latest/module_file_support.html#exclude-or-include-specific-module-files). We'll add that to default config soon.

**Comment by jcwright77** [comment]
I did that manually but it doesn't help. The main problem seems to be 
compiler dependencies in the modules.

eg

core compiler gcc/8.5.0

library built with gcc/12.2.0 and the library is loaded via module

now load intel compiler which was built with gcc/8.5.0 since it is a 
compiler and gcc/8.5.0 is the core compiler

msg:

Lmod is automatically replacing "gcc/12.2.0" with 
"intel-oneapi-compilers/2024.1.0"

because both are in the compiler family.

removing the glibc dependency doesn't help (same version in both modules 
anyway)

So the problem remains: can one use a gcc built library to build a 
program (out of band, eg not in space) with another compiler ,eg intel? 
This is very common outside of spack. Maybe I could require intel to be 
built with gcc/12.2.0 but then it wouldn't be under 'Core' in lmod.

On 2024-10-30 05:17, Harmen Stoppels wrote:

> Exclude glibc from the modules. We'll add that to default config soon.
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe 
> [2].
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>

-- 
-john
Principal Research Scientist John Wright
Office 617-253-9612
zoom: https://mit.zoom.us/my/jcwright



Links:
------
[1] https://github.com/spack/spack/issues/47319#issuecomment-2446279622
[2]

**Comment by haampie** [comment]
I don't immediately get what the relevance of `gcc/8.5.0` is, it's not related to any error right?

Is there an actual error? Notice you could remove `family(...)` statements from compiler modules.

Note that this is a self-imposed constraint in LMod's model, which is somewhat detached from reality:

https://lmod.readthedocs.io/en/latest/080_hierarchy.html

> Libraries built with one compiler need to be linked with applications built with the same compiler version.

**Comment by jcwright77** [comment]
Yes, I can understand the module constraint. Of course in spack concretization this is not necessarily enforced for example system libraries, compiled by c can be used in Intel. or in outside of Spack builds that is very common. I think there are two ways around this one is to make an environment, and then load  the environment with a module for outside of spack compilations. The other is to actually make a spack package for the final target. Thanks. We can close this.   -johnOn Oct 30, 2024, at 12:28 PM, Harmen Stoppels ***@***.***> wrote:﻿
Is it an error though? You could consider removing family(...) statements.
Note that this is a self-imposed constraint by LMod which is detached from reality:
https://lmod.readthedocs.io/en/latest/080_hierarchy.html

Libraries built with one compiler need to be linked with applications built with the same compiler version.

is how it starts...
